pkg/cosign/verify.go

```diff
@@ -235,55 +247,125 @@ func Verify(ctx context.Context, signedImgRef name.Reference, accessor Accessor,
 			bundleVerified = bundleVerified || verified
 
 			if !verified && co.RekorURL != "" {
-				// Get the right public key to use (key or cert)
-				var pemBytes []byte
 				if co.SigVerifier != nil {
-					var pub crypto.PublicKey
-					pub, err = co.SigVerifier.PublicKey(co.PKOpts...)
+					pub, err := co.SigVerifier.PublicKey(co.PKOpts...)
 					if err != nil {
 						return err
 					}
-					pemBytes, err = cryptoutils.MarshalPublicKeyToPEM(pub)
-				} else {
-					pemBytes, err = cryptoutils.MarshalCertificateToPEM(cert)
+					return tlogValidatePublicKey(rekorClient, pub, b64sig, payload)
 				}
+
+				return tlogValidateCertificate(rekorClient, cert, b64sig, payload)
+			}
+			return nil
+		}(sig); err != nil {
+			validationErrs = append(validationErrs, err.Error())
+			continue
+		}
+
+		// Phew, we made it.
+		checkedSignatures = append(checkedSignatures, sig)
+	}
+	if len(checkedSignatures) == 0 {
+		return nil, false, fmt.Errorf("no matching signatures:\n%s", strings.Join(validationErrs, "\n "))
+	}
+	return checkedSignatures, bundleVerified, nil
+}
+
+// VerifyAttestations does all the main cosign checks in a loop, returning the verified attestations.
+// If there were no valid attestations, we return an error.
+func VerifyImageAttestations(ctx context.Context, signedImgRef name.Reference, co *CheckOpts) (checkedAttestations []oci.Signature, bundleVerified bool, err error) {
+	// Enforce this up front.
+	if co.RootCerts == nil && co.SigVerifier == nil {
+		return nil, false, errors.New("one of verifier or root certs is required")
+	}
+
+	// TODO(mattmoor): We could implement recursive verification if we just wrapped
+	// most of the logic below here in a call to mutate.Map
+
+	se, h, err := getSignedEntity(signedImgRef, co.RegistryClientOpts)
+	if err != nil {
+		return nil, false, err
+	}
+	atts, err := se.Attestations()
+	if err != nil {
+		return nil, false, err
+	}
+	sl, err := atts.Get()
+	if err != nil {
+		return nil, false, err
+	}
+
+	validationErrs := []string{}
+
+	var rekorClient *client.Rekor
+	if co.RekorURL != "" {
+		rekorClient, err = rekor.GetRekorClient(co.RekorURL)
+		if err != nil {
+			return nil, false, err
+		}
+	}
+	for _, att := range sl {
+		if err := func(sig oci.Signature) error {
+			b64sig, payload, cert, err := unpackOCISignature(att)
+			if err != nil {
+				return err
+			}
+
+			verifier := co.SigVerifier
+			if verifier == nil {
+				// If we don't have a public key to check against, we can try a root cert.
+				if cert == nil {
+					return errors.New("no certificate found on attestation")
+				}
+				verifier, err = validateAndUnpackCert(cert, co)
 				if err != nil {
 					return err
 				}
+			}
+			verifier = newReverseDSSEVerifier(verifier)
 
-				// Find the uuid then the entry.
-				uuid, _, err := FindTlogEntry(rekorClient, b64sig, payload, pemBytes)
-				if err != nil {
+			if err := verifyBase64Signature(ctx, verifier, b64sig, payload); err != nil {
```

Comment:
Rather than passing b64sig, which we know to be `""` call this `verifyAttestation`, push the `newReverseDSSEVerifier` wrapper into this call, and have this take the signature.